src/derived.js

```diff
@@ -49,3 +50,72 @@ export const diffClamp = function(a, minVal, maxVal) {
     min(max(add(cond(defined(value), value, a), diff(a)), minVal), maxVal)
   );
 };
+
+const interpolateInternalSingle = function(
+  value,
+  inputRange,
+  outputRange,
+  offset
+) {
+  const inS = inputRange[offset];
+  const inE = inputRange[offset + 1];
+  const outS = outputRange[offset];
+  const outE = outputRange[offset + 1];
+  const progress = divide(sub(value, inS), sub(inE, inS));
+  return add(outS, multiply(progress, sub(outE, outS)));
+};
+
+const interpolateInternal = function(
+  value,
+  inputRange,
+  outputRange,
+  offset = 0
+) {
+  if (inputRange.length - offset === 2) {
+    return interpolateInternalSingle(value, inputRange, outputRange, offset);
+  }
+  return cond(
+    lessThan(value, inputRange[offset + 1]),
+    interpolateInternalSingle(value, inputRange, outputRange, offset),
+    interpolateInternal(value, inputRange, outputRange, offset + 1)
+  );
+};
+
+export const Extrapolate = {
+  EXTEND: 'EXTEND',
+  CLAMP: 'CLAMP',
+  IDENTITY: 'IDENTITY',
+};
+
+export const interpolate = function(value, config) {
+  const {
+    inputRange,
+    outputRange,
+    extrapolate = Extrapolate.EXTEND,
+    extrapolateLeft,
+    extrapolateRight,
+  } = config;
+  const left = extrapolateLeft || extrapolate;
+  const right = extrapolateRight || extrapolate;
+  let output = interpolateInternal(value, inputRange, outputRange);
+
+  if (left === Extrapolate.EXTEND) {
+  } else if (left === Extrapolate.CLAMP) {
+    output = max(outputRange[0], output);
```

Comment:
not entirely sure but this seems to be only correct if outputRange was sorted from min to max which is not a requirement. We only expect inputRange to be sorted that way.

I think it would be better to do something along these lines both for CLAMP and IDENTITY:
```js
if (left === Extrapolate.EXTEND) {
} else if (left === Extrapolate.CLAMP) {
  output = cond(lessThan(value, inputRange[0]), outputRange[0], output);
} else if (left === Extrapolate.IDENTITY) {
  output = cond(lessThan(value, inputRange[0]), value, output);
}
```

similarly for "right"